common/flatpak-utils.c

```diff
@@ -3665,6 +3669,8 @@ flatpak_repo_update (OstreeRepo   *repo,
 
       /* Add bindings to the metadata. */
       new_summary_commit_dict = g_variant_dict_new (new_summary);
+      g_variant_dict_insert_value (new_summary_commit_dict, "xa.commits",
+                                   g_variant_builder_end (&commits_builder));
```

Comment:
I was sad to find out that `ostree-metadata` doesn't contain the refs mapping already. I think this is needed, but perhaps we can do this in ostree? I.e., instead of rolling the `ostree-metadata` code in both the ostree and flatpak summary generation CLI commands it can be done in the `regenerate_summary` API if the repo has a collection ID. Then augment it to always include the collection refs map as in the summary.

@pwithnall I was under the impression that `ostree-metadata` was intended to be a replacement for the summary file. Is there a reason it doesn't have the collection refs map in it?

Reply:
> I was under the impression that `ostree-metadata` was intended to be a replacement for the summary file. Is there a reason it doesn't have the collection refs map in it?

It’s not meant to be a direct replacement. You still use the (unsigned) `summary` file for the collection–refs → commit checksums map, and then verify them using the signed binding data in the commit metadata.

Reply:
That means you have to pull each commit object for verification in the p2p case, which might require authorization for purchased refs. That means that the refs can't be reliably resolved unless the user is authenticated, which is what @alexlarsson is trying to avoid.

Do you have any objections if the collection map is included in `ostree-metadata`?

Reply:
The ostree-metadata data is mostly from flatpak anyway (ostree only sets two properties there) so its not a huge difference. 

So, the p2p summary file *does* have the commit id in it, but that is the commit id of the ref as stored in the p2p repo. What this commit adds is the commit id of the cached ref flatpak metadata in the ostree-metadata ref, which may be different from the commit id of the actual ref.

Reply:
> That means you have to pull each commit object for verification in the p2p case, which might require authorization for purchased refs. That means that the refs can't be reliably resolved unless the user is authenticated, which is what @alexlarsson is trying to avoid.

Do they have to be *reliably* resolved until you’re actually ready to go through the authenticator?

> Do you have any objections if the collection map is included in `ostree-metadata`?

Yes, some mild ones: (a) it increases the size of `ostree-metadata` and means it has to be updated *much* more frequently than it currently is; (b) the collection map still needs to be in the `summary` file too, because otherwise it’s not possible to resolve `ostree-metadata` itself. Having it in two places, or parts of it in two places, is ugly and confusing.

> What this commit adds is the commit id of the cached ref flatpak metadata in the ostree-metadata ref, which may be different from the commit id of the actual ref.

Sorry, I don’t understand. What do you mean by ‘cached ref’ and ‘actual ref’?

Reply:
> > That means you have to pull each commit object for verification in the p2p case, which might require authorization for purchased refs. That means that the refs can't be reliably resolved unless the user is authenticated, which is what @alexlarsson is trying to avoid.
> 
> Do they have to be _reliably_ resolved until you’re actually ready to go through the authenticator?

The commit object contains the metadata for the app, which says which runtime we need to download (this is what resolve means in the scope of a flatpak transaction). If the ostree-metadata is mismatched then the resolve would add a download operation for a runtime that will not match what will actually be installed. 

I'm not sure if this answers your question, because I'm not sure what you mean. But we definitely need to actually download the commit to get the right deps, and for that we need the token. We don't necessarily have to have the most up-to-date info to just *know* that we need to get a token though, which is what the iterative workaround uses.

> > Do you have any objections if the collection map is included in `ostree-metadata`?
> 
> Yes, some mild ones: (a) it increases the size of `ostree-metadata` and means it has to be updated _much_ more frequently than it currently is; (b) the collection map still needs to be in the `summary` file too, because otherwise it’s not possible to resolve `ostree-metadata` itself. Having it in two places, or parts of it in two places, is ugly and confusing.
>
> > What this commit adds is the commit id of the cached ref flatpak metadata in the ostree-metadata ref, which may be different from the commit id of the actual ref.
> 
> Sorry, I don’t understand. What do you mean by ‘cached ref’ and ‘actual ref’?

So, both these questions sound confusing. Lemme back up a bit.

Both the master and the p2p repo have a summary file. It contains a mapping from ref name to the commit name that is available (as HEAD) in each individual repo.

Additionally, the summary in the master repo contains a cache (xa.cache) for information about flatpak apps in the repo that is used for resolving. For example, its used when downloading an app to know which runtime to download. Additionally it has some "nice to have" data like installed and download size. These caches don't actually exists for *all* refs, just flatpak ones. xa.cache doesn't contain the commit id of the version we extracted the data from, because being inside the summary we can assume the that the cache and the regular ref->commit maps are synced, so this data is already available.

However, in the p2p case each peer generates the summary file, so it is not trustworthy for critical information (i.e. not signed). So the master also makes a copy of xa.cache inside the commit of the ostree-metadata ref, which *is* signed, and this is mirrored as-is to the p2p nodes. However, at this point we disconnected the cached metadata from the summary so we can't rely on it to know exactly what commit each cached metadata came from. In order to fix this the PR adds an array that matches the order of xa.cache that just contains the commit that the cached data refers to.

Now to answer:
(a) yes, it slightly increases the size. Compared to the size of the rest of the cached data already in it this is pretty insignificant thought.
(b) The collection map (ref->commit id) is not quite the same as the "what commit did we cache metadata from" map, and will not be used during an ostree pull. It also would never include ostree-metadata or any non-flatpak refs.

cached ref: The version of the ref that we extracted data from and put in the xa.cache.
actual ref: The version of the ref that is currently available in the p2p mirror.

In a full mirror, or on the master these would be the same, but in a partial mirror where the ostree-metadata branch and some other ref were mirrored at different times these would not be the same.

Reply:
> In order to fix this the PR adds an array that matches the order of xa.cache that just contains the commit that the cached data refers to.

That’s the bit of information I’d missed, thanks. This makes sense now, and I can’t immediately think of a better approach, assuming you can’t break backwards compatibility with the `xa.cache` format and change it to map from checksum → cache data, rather than ref name → cache data.

Reply:
> > In order to fix this the PR adds an array that matches the order of xa.cache that just contains the commit that the cached data refers to.
> 
> That’s the bit of information I’d missed, thanks. This makes sense now, and I can’t immediately think of a better approach, assuming you can’t break backwards compatibility with the `xa.cache` format and change it to map from checksum → cache data, rather than ref name → cache data.

This still feels like a general ostree p2p issue to me, though, since you can't trust the ref map in the peer's summary. If you include the ref map also in the ostree-metadata commit, then you can validate the peer's mirrored ref without actually pulling it. The flow would be:

* Fetch the summary from the peer
* Pull refs/mirrors/<id>/ostree-metadata and verify it against the remote's trusted key
* Verify that the commits pointed to by the collection map in the summary match the ref map in the mirrored ostree-metadata commit
* Use the ref data and additional metadata (such as `xa.cache`) in the mirrored ostree-metadata, filtering the refs to those in the summary collection map

Then all you're getting from the peer is a list of which refs it has. The rest of the information comes from the remote and is verified. You can the carry on pulling mirrored commits with certainty that they came from the remote.

This is the essentially the same as what this commit is doing but rather in ostree. And, if the full ref map is included, then you can get the verified commit timestamp from the remote as well to determine if it's older than what you already have.

Reply:
FWIW, I implemented that in ostreedev/ostree#1946. I do believe that's the right thing to do, but I've spent much less time on P2P than on other parts of ostree.

Reply:
I don’t think you can reliably do that, since the peer might update some refs from a collection out of sync with other refs from the same collection (say, they update from a USB stick which has some, but not all, apps). At that point, whatever `ostree-metadata` commit they have for that collection (new or old) is going to have a mapping of collection–ref to checksum which doesn’t match some of the refs in the peer’s repository.

That’s why I think the verification of “is it OK for collection–ref R to point to commit C” needs to be done via commit bindings, since the verification data (the binding) is guaranteed to be in sync with the object you’re verifying (the commit).

I see two ways of dealing with this:
 1. Operate on commit checksums, rather than collection–refs, everywhere except in the initial mapping from collection–ref to checksum (which happens using the map from the `summary` file). Then the metadata for several versions (commits) of an app can be carried around without conflicting, and peers use the data for whichever commit they have for a given app.
 2. Be OK with (getting a token and) downloading commit objects which we might then throw away because their binding data doesn’t validate.

Reply:
Yeah, i don't believe the change in https://github.com/ostreedev/ostree/pull/1946 is really useful/correct, and has commented so there.

Reply:
Basically, from a practical point of view, its not going to be guaranteed that all the refs in a p2p peer were downloaded at the same time, but requiring them to match the ostree-meta from upstream requires that. And anyway, we can just verify the signatures and binding of the commit objects we download, so why would we need to verify the commits themselves against the ostree-metadata.

Now, obviously this is for the refs themselves, for the data *in* the ostree-meta commit we do need to know what commits it refers to, but the solution in this PR seems more efficient and clean for achieving that.